### PR TITLE
Add shared gRPC mode to run the native provider

### PR DESCRIPTION
**Pull request by ulucinar**
### Description of your changes

Fixes https://github.com/crossplane/terrajet/issues/261

Forks the native Terraform plugin as a shared server if `config.Provider.SharedGRPC` is set. Depends on https://github.com/crossplane/terrajet/pull/267.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

[contribution process]: https://git.io/fj2m9
Tested with https://github.com/crossplane/terrajet/pull/267

### CODE REVIEW

File: cmd/provider/main.go [hunk 1]
```diff
@@ -21,18 +21,18 @@ import (
 	"path/filepath"
 	"time"
 
-	"github.com/crossplane/crossplane-runtime/pkg/feature"
-	tjcontroller "github.com/crossplane/terrajet/pkg/controller"
-	"k8s.io/client-go/tools/leaderelection/resourcelock"
-
 	xpcontroller "github.com/crossplane/crossplane-runtime/pkg/controller"
```
Comment (muvaf) [inline cmd/provider/main.go:24]: go mod ordering

File: internal/clients/template.go [hunk 2]
```diff
@@ -86,15 +74,19 @@ func TerraformSetupBuilder(version, providerSource, providerVersion string) terr
...
+		// credentials via the Terraform main.tf.json instead.
+		/*ps.Env = []string{
+			fmt.Sprintf("%s=%s", "HASHICUPS_USERNAME", templateCreds["username"]),
+			fmt.Sprintf("%s=%s", "HASHICUPS_PASSWORD", templateCreds["password"]),
+		}*/
+		// set credentials in Terraform provider configuration
+		/*ps.Configuration = map[string]interface{}{
+			"username": templateCreds["username"],
+			"password": templateCreds["password"],
+		}*/
```
Comment (muvaf) [inline internal/clients/template.go:89]:
```suggestion
		ps.Configuration = map[string]interface{}{
			"username": templateCreds["username"],
			"password": templateCreds["password"],
		}
```

I think it's fine to have this uncommented to keep in working state by default.

File: cmd/provider/main.go [hunk 3]
```diff
@@ -84,8 +84,12 @@ func main() {
 			Features:                &feature.Flags{},
 		},
 		Provider:       config.GetProvider(),
-		WorkspaceStore: terraform.NewWorkspaceStore(log),
-		SetupFn:        clients.TerraformSetupBuilder(*terraformVersion, *providerSource, *providerVersion),
+		WorkspaceStore: terraform.NewWorkspaceStore(log), // uncomment the following two WorkspaceStore configuration
```
Comment (muvaf) [inline cmd/provider/main.go:87]: Needs update

**Review by muvaf — COMMENTED** [review]
Thanks @ulucinar !

File: cmd/provider/main.go (hunk 3 above)
Comment (ulucinar) [reply cmd/provider/main.go]: Done.

File: internal/clients/template.go (hunk 2 above)
Comment (ulucinar) [reply internal/clients/template.go:89]: [null provider](https://registry.terraform.io/providers/hashicorp/null/latest/docs) does not have any configuration options.

File: cmd/provider/main.go (hunk 1 above)
Comment (ulucinar) [reply cmd/provider/main.go]: Done.

File: Makefile [hunk 4]
```diff
@@ -9,7 +9,8 @@ export TERRAFORM_VERSION := 1.1.6
 export TERRAFORM_PROVIDER_SOURCE := hashicorp/null
 export TERRAFORM_PROVIDER_VERSION := 3.1.0
 export TERRAFORM_PROVIDER_DOWNLOAD_NAME := terraform-provider-null
-export TERRAFORM_PROVIDER_DOWNLOAD_URL_PREFIX := https://releases.hashicorp.com/terraform-provider-null/3.1.0
+export TERRAFORM_PROVIDER_DOWNLOAD_URL_PREFIX := https://releases.hashicorp.com/${TERRAFORM_PROVIDER_DOWNLOAD_NAME}/${TERRAFORM_PROVIDER_VERSION}/${TERRAFORM_PROVIDER_DOWNLOAD_NAME}_${TERRAFORM_PROVIDER_VERSION}_linux
+export TERRAFORM_NATIVE_PROVIDER_BINARY := ${TERRAFORM_PROVIDER_DOWNLOAD_NAME}_v${TERRAFORM_PROVIDER_VERSION}_x5
```
[the remark this answers is not in the document]
Comment (ulucinar) [reply Makefile]: Done.

**Review by muvaf — APPROVED** [review]
Thanks @ulucinar !

**Pull request merged**